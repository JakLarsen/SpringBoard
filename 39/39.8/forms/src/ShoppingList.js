import React, {useState} from "react";      
import Item from "./Item";
import ProductForm from "./ProductForm";
import { v4 as uuidv4, v4 } from 'uuid';





const ShoppingList = () => {
    const INITIAL_STATE = [
        {
           id: uuidv4(),
           name: "Peanut Butter",
           qty: 2 
        },
        {
            id: uuidv4(),
            name: "Whole Milk",
            qty: 1
        }
    ]
    const [items, setItems] = useState(INITIAL_STATE)

    const addItemToShoppingList = (newItem) => {
        setItems(items => [...items, {...newItem, id: uuidv4()}])
    }

    return (
        <div>
            <h3>Shopping List</h3> 
            <div>
                {items.map(item=>
                    <Item name={item.name} qty={item.qty} key={item.id} id={item.id}/>
                )}
            </div>
            <ProductForm addItemToShoppingList={addItemToShoppingList}/>

        </div>
    )


}

export default ShoppingList