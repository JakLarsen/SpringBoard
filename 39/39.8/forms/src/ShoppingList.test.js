import React from 'react';
import {render, fireEvent} from "@testing-library/react"
import ShoppingList from './ShoppingList';


it('Renders without crashing', function(){
    render(<ShoppingList/>)
});

it('Matches snapshot', function(){
    const {asFragment} = render(<ShoppingList/>)
    expect(asFragment()).toMatchSnapshot()
});

it('Should add new item: Name and Quantity inputs function properly', function() {
    //Get queryByText and getByLabelText
    const {queryByText, getByLabelText} = render(<ShoppingList/>)
    //Find our input from our label text
    const input = getByLabelText('Name')
    const qtyInput = getByLabelText('Quantity')
    //Find our submit btn by our btn text
    const btn = queryByText('Done')
    //Expect 'Product: Chocolate Milk' to not be found in the document yet
    expect(queryByText('Product: Chocolate Milk')).not.toBeInTheDocument();
    expect(queryByText('Amount: 7')).not.toBeInTheDocument();
    //Give our input a value of 'Chocolate Milk'
    fireEvent.change(input, {target: {value: 'Chocolate Milk'}})
    fireEvent.change(qtyInput, {target: {value: 7}})
    //Submit our form
    fireEvent.click(btn)
    //Now expect our full label + value product to have be added to the dom
    expect(queryByText('Product: Chocolate Milk')).toBeInTheDocument();
    expect(queryByText('Amount: 7')).toBeInTheDocument();

});
