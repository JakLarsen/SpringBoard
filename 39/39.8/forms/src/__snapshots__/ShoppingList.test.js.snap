// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Shopping List
    </h3>
    <div>
      <div>
        <h3>
          Product: Peanut Butter
        </h3>
        <h4>
          Amount: 2
        </h4>
      </div>
      <div>
        <h3>
          Product: Whole Milk
        </h3>
        <h4>
          Amount: 1
        </h4>
      </div>
    </div>
    <form>
      <label
        for="name"
      >
        Name
      </label>
      <input
        id="name"
        name="name"
        placeholder="product"
        type="text"
        value=""
      />
      <label
        for="qty"
      >
        Quantity
      </label>
      <input
        id="qty"
        max="10"
        min="1"
        name="qty"
        placeholder="amount"
        type="range"
        value="0"
      />
      <h4>
        0
      </h4>
      <button>
        Done
      </button>
    </form>
  </div>
</DocumentFragment>
`;
